### PR TITLE
Build variant support [ESD-1085] [v2.3.0]

### PR DESCRIPTION
**Pull request by silverjam**
Backport of #1242 to v2.3.0

### CODE REVIEW

**Comment by benjaminaltieri** [comment]
also do we need https://github.com/swift-nav/piksi-releases/pull/94 to get approved/merged before we can get tags working? is there a safe way to validate that these integrate well?

File: scripts/list-public-artifacts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+  while read -r line; do
+
+    repo=$(field 1 "$line")
+    version=$(field 2 "$line")
+    object=$(field 3 "$line")
+
+    echo "${repo//'{version}'/} ${version} ${object}"
+
+  done < <(jq -r "${q_artifacts}|${f_upgrade_tool}|$t_pub_artifacts" "$d")
```
[the remark this answers is not in the document]
Comment (silverjam) [reply scripts/list-public-artifacts]: looks correct, the only object that needs to be copied to `swiftnav-releases` is piksi_upgrade_tool

**Comment by silverjam** [comment]
@benjaminaltieri Actually need to merge #1242 and #1246 prior to merging the piksi-releases changes

File: Makefile [hunk 2]
```diff
@@ -7,158 +7,48 @@ ifeq ($(HW_CONFIG),)
...
+		\( ! -path buildroot/$(VARIANT_OUTPUT)/images \) \
 		-print -exec rm -rf {} \;
-	rm -rf buildroot/output/images/*
+	rm -rf buildroot/$(VARIANT_OUTPUT)/images/*
 
 flush-rootfs:
-	find buildroot/output -name .stamp_target_installed -delete
-	rm -rf buildroot/output/target/*
-
-# 'Package-specific:'
```
[the remark this answers is not in the document]
Comment (silverjam) [reply Makefile]: didn't mean to delete these comments actually

File: scripts/fetch-external-artifacts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    for artifact in artifact_set:
+        ref_sha256 = PBR.get_artifact_prop(artifact, 'sha256')
+        local_path = PBR.get_artifact_prop(artifact, 'local_path')
+        if os.path.exists(local_path) and ref_sha256 == PBR.sha256digest(local_path):
+            print('>>> SHA256 matched (no download needed): {}'
+                  .format(local_path))
+            continue
+        s3_url = PBR.build_s3_artifact_url(artifact)
+        print('>>> Downloading: {}'.format(s3_url))
+        R().call('aws', 's3', 'cp', s3_url, local_path).check()
```
[the remark this answers is not in the document]
Comment (silverjam) [reply scripts/fetch-external-artifacts]: Yes, added

File: scripts/yaml2json [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    if platform.system() == 'Darwin':
+        system = 'darwin'
+    else:
+        system = 'linux'
+    bin_name = 'yaml2json_{}_{}'.format(system, arch)
+    path = OP.join(OP.dirname(OP.realpath(__file__)), 'yaml2json.bin', bin_name)
+    return path
+
+
+def _run_yaml2json():
```
[the remark this answers is not in the document]
Comment (silverjam) [reply scripts/yaml2json]: De-duped

File: board/piksiv3/post_image.sh [hunk 5]
```diff
@@ -42,10 +48,12 @@ SDK_FPGA_SHA1="${BR2_EXTERNAL_piksi_buildroot_PATH}/firmware/prod/piksi_sdk_fpga
 
 DEV_BIN_PATH=$OUTPUT_DIR/PiksiMulti-DEV-$FILE_GIT_STRING.bin
 FAILSAFE_BIN_PATH=$OUTPUT_DIR/PiksiMulti-FAILSAFE-$FILE_GIT_STRING.bin
+
+IMAGE_NAME=$("$SCRIPTS/get-variant-prop" "$VARIANT" image_name version=$FILE_GIT_STRING)
+IMAGE_BIN_PATH=$OUTPUT_DIR/$IMAGE_NAME
+
 INTERNAL_BIN_PATH=$OUTPUT_DIR/PiksiMulti-INTERNAL-$FILE_GIT_STRING.bin
```
[the remark this answers is not in the document]
Comment (silverjam) [reply board/piksiv3/post_image.sh]: removed

File: scripts/list-public-artifacts [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+yaml2json <"external-artifacts.yaml" >"$d"
+
+q_artifacts='.artifact_sets[]|.artifact_set.artifacts[]'
+t_pub_artifacts='"\(.s3_repository) \(.version) \(.s3_object)"'
+f_upgrade_tool='select(.name=="piksi_upgrade_tool")'
+
+field() { fnum=$1; shift; echo "$@" | awk "{print \$$fnum}"; }
+
+gen_s3_pathes() {
```
[the remark this answers is not in the document]
Comment (silverjam) [reply scripts/list-public-artifacts]: 🔨

File: external-artifacts.yaml [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    artifacts:
+    - name: rtos_elf
+      version: v2.2.0-develop-2019050321
+      s3_bucket: swiftnav-artifacts
+      s3_repository: piksi_firmware_private
+      s3_object: v3/piksi_firmware_v3_prod.stripped.elf
+      local_path: firmware/prod/piksi_firmware.elf
+      sha256: 089e69ac16ca3c27c7ede78e30366775d943473402d9a066233009a51e3dbf5e
+    - name: fpga_bitstream
+      version: v2.2.0-develop-2019050321
```
[the remark this answers is not in the document]
Comment (silverjam) [reply external-artifacts.yaml]: I need to be careful moving these changes between release and master, this in actually an external artifact for master, fixed

**Comment by silverjam** [comment]
Previous error was on the SDK artifact hash (https://jenkins.ci.swift-nav.com/blue/organizations/jenkins/swift-nav%2Fpiksi_buildroot/detail/PR-1246/24/pipeline), fixed that, but reintroduced "when" guard to prevent SDK from building on non-tag pushes so we won't get an indication that this worked until after a tag build... merging.

**Pull request merged**